File: Algo/Leetcode/023MergeKSortedLists.cpp
//1. Compare the head of the k lists. Then recurrsion. Remember to consider nulls. O(kn).
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0) return NULL;
        ListNode* res = nullptr;
        if(!lists[0]) {lists.erase(lists.begin());return mergeKLists(lists);}
        int minN = lists[0]->val;
        int minId = 0;
        for(int i=1;i<lists.size();i++){
            if(!lists[i]){
                continue;
            }
            if(lists[i]->val < minN){
                minN = lists[i]->val;
                minId = i;
            }
        }
      
        res = lists[minId];
        if(!lists[minId] -> next){
            lists.erase(lists.begin()+minId);
        }
        else lists[minId] = lists[minId] -> next;
        res -> next = mergeKLists(lists);
        return res;
    }
};

// 2. 两两merge。  O(nlogk).
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0) return NULL;
        ListNode* res = NULL;
        int k = lists.size();
        int idx;
        while(k>1){
            idx = 0;
            for(int i=0; i<k; i+=2){
                if(i==k-1) lists[idx++] = lists[i];
                else lists[idx++] = merge2Lists(lists[i], lists[i+1]);
            }
            k = idx;
        }
        return lists[0];
    }

    ListNode* merge2Lists(ListNode* a, ListNode* b){
        if(!a && !b) return NULL;
        if(!a) return b;
        if(!b) return a;

        ListNode* res;
        if(a->val < b->val){
            res = a;
            a = a->next;
        } else{
            res = b;
            b = b->next;
        }
        res -> next = merge2Lists(a, b);
        return res;
    }
};

File: Algo/Leetcode/023MergeKSortedLists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        k = len(lists)
        if k == 0:
            return None
        while k > 1:
            _lists = []
            if k % 2 == 0:
                i = 0
                while i < k:
                    _lists.append(self.merge2Lists(lists[i], lists[i+1]))
                    i += 2
                k = k / 2
            else:
                i = 0
                while i < k-1:
                    _lists.append(self.merge2Lists(lists[i], lists[i+1]))
                    i += 2
                _lists.append(lists[i])
                k = (k + 1) / 2

            lists = _lists

        return lists[0]

    def merge2Lists(self, list1, list2):
        head = ListNode()
        res = head
        while list1 or list2:
            if list1 and list2:
                if list1.val <= list2.val:
                    head.next = ListNode(list1.val)
                    list1 = list1.next
                else:
                    head.next = ListNode(list2.val)
                    list2 = list2.next
            elif list1 is None:
                head.next = ListNode(list2.val)
                list2 = list2.next
            else:
                head.next = ListNode(list1.val)
                list1 = list1.next

            head = head.next

        return res.next
